README.md
# Blinkbox::CommonLogging

Automatic configuration for logging in the blinkbox format.

## Usage

```ruby
# $ gem install blinkbox-common_logging
require "blinkbox/common_logging"

log_server = {
  host: "127.0.0.1",
  port: 12201,
  # This is the name of the invoked script
  facility: $0,

}
# These are the defaults

logger = Blinkbox::CommonLogging.new(log_server)

# Usual levels are available
logger.debug "This is a basic debug message"
logger.info  "This is a basic info message"
logger.warn  "This is a basic warn message"
logger.error "This is a basic error message"
logger.fatal "This is a basic fatal message"

# You can also pass an exception to #notify
begin
  "raise an exception".by_calling_a_missing_method
raise StandardError => e
  logger.notify(e)
end
```

CHANGELOG.md
# Change log

## 0.5.2 (2015-01-29 12:13:21)

Unknown graylog server hostname fix

### Bug fixes

- Monkey patch to allow execution of code when the given graylog server name is unresolvable.

## Open Source release (2015-01-28 14:11:21)

Today we have decided to publish parts of our codebase on Github under the [MIT licence](LICENCE). The licence is very permissive, but we will always appreciate hearing if and where our work is used!

## 0.5.1 ([#8](https://git.mobcastdev.com/Platform/common_logging.rb/pull/8) 2014-12-15 15:01:11)

Changed level mapping to :direct

### Bug fixes

- Changed the level mapping to be `:direct` which means that
`logger.error` causes an error level message in Graylog, `logger.warn`
causes a warning level message, and so on. The default mappings mean
that `logger.error` causes a warning level message, and `logger.warn`
causes a notice level message which is nuts.

## 0.5.0 ([#7](https://git.mobcastdev.com/Platform/common_logging.rb/pull/7) 2014-10-03 15:02:14)

Log to console as well as Graylog

### New features

- Setting `logging.console.enabled` to `true` will now output to stdout as well as graylog.
- Slight refactoring of code in files.

## 0.4.1 ([#6](https://git.mobcastdev.com/Platform/common_logging.rb/pull/6) 2014-10-02 08:31:38)

Validation of values

### Improvements

- Validates input from config and raises an `ArgumentError` if anything is unusable.
- Will list out *all* problematic variables so fixing the config is faster.
- Translates the named invalid keys between `from_config` and `new` so that the error message makes sense for the code being used.

## 0.4.0 ([#5](https://git.mobcastdev.com/Platform/common_logging.rb/pull/5) 2014-09-25 10:46:32)

Shallow hashes

### New feature

- Graylog can't deal with nested hashes, so I've added functionality to flatten deep hashes given to the logging methods into dot-separated shallow ones:

```ruby
{
  a: {
    b: "will become a.b"
  }
}
```

becomes

```ruby
{
  "a.b" => "will become a.b"
}
```

## 0.3.1 ([#4](https://git.mobcastdev.com/Platform/common_logging.rb/pull/4) 2014-09-22 13:53:00)

Ensure from_config actually works from a config!

### Bugfix

- `.from_config` now expects a shallow hash (with dots in keys) just as common_config.rb provides.

## 0.3.0 ([#3](https://git.mobcastdev.com/Platform/common_logging.rb/pull/3) 2014-08-14 16:45:39)

Allow adding of service version to default params

### New feature

- Allows us to add the default `serviceVersion` parameter to the messages sent to GrayLog which, along with the facility (which should be the service name) will allow us to pinpoint what code needs to be looked at when issues occur.

## 0.2.0 ([#2](https://git.mobcastdev.com/Platform/common_logging.rb/pull/2) 2014-08-14 13:09:41)

Support loading directly from config

### New feature

- Allow the injection of a hash from common_config to initialise a logger.

## 0.1.0 ([#1](https://git.mobcastdev.com/Platform/common_logging.rb/pull/1) 2014-08-12 17:05:06)

First release

### New Features

- Logs messages and exceptions.
- Pretty much only wraps the `GELF::Notifier` class (may need to check log levels) at the moment, but we may extend this to include extra functionality as the ruby logging requirements become clear.



spec/spec_helper.rb
$LOAD_PATH.unshift File.join(__dir__, "../lib")
require "blinkbox/common_logging"
require "rspec/mocks"

module Helpers
  def expected_hash(hash)
    n = Blinkbox::CommonLogging.new
    n.send(:extract_hash, hash)
    n.send(:serialize_hash)
    hash = n.instance_variable_get(:'@hash')

    # The following will be different in test code, so we just ignore them
    hash.delete_if { |key, v| %w{timestamp line file}.include?(key) }
  end
end

class GELF::Notifier
  def self.last_hash
    # The following will be different in test code, so we just ignore them
    @@last_hash.delete_if { |key, v| %w{timestamp line file}.include?(key) }
  end

  private
  alias :old_serialize_hash :serialize_hash
  def serialize_hash
    @@last_hash = @hash
    old_serialize_hash
  end
end

RSpec.configure do |c|
  c.include Helpers
end

spec/common_logging_spec.rb
context Blinkbox::CommonLogging do
  describe "logger interface" do
    before :each do
      fake_class = class_double("GELF::RubyUdpSender")
      stub_const("GELF::RubyUdpSender", fake_class)
      @fake = instance_double("GELF::RubyUdpSender")
      allow(fake_class).to receive(:new).and_return(@fake)
      allow(@fake).to receive(:send_datagrams)
    end

    subject(:logger) { described_class.new }

    %i{debug info warn error fatal}.each do |level|
      it "must deliver #{level} level string messages to GrayLog" do
        log_string = "Test #{level} string"
        logger.send(level, log_string)
        
        log_hash = expected_hash({
          short_message: log_string,
          level: GELF.const_get(level.to_s.upcase)
        })
        expect(GELF::Notifier.last_hash).to eq(log_hash)
      end

      it "must also log #{level} messages to stdout if if enabled" do
        log_string = "Test #{level} string with console output"
        log_hash = expected_hash({
          short_message: log_string,
          level: GELF.const_get(level.to_s.upcase)
        })
        common_logger = described_class.new(echo_to_console: true)
        stdout_logger = common_logger.instance_variable_get(:'@stdout_logger')
        expect(stdout_logger).to be_a(Logger)
        stdout_logger = instance_double(Logger)
        allow(stdout_logger).to receive(level.to_sym)
        common_logger.instance_variable_set(:'@stdout_logger', stdout_logger)
        common_logger.send(level, log_string)
        expect(GELF::Notifier.last_hash).to eq(log_hash)
        expect(stdout_logger).to have_received(level.to_sym).with(hash_including("short_message" => log_string))
      end
    end

    it "must log exceptions at the error level" do
      begin
        raise RuntimeError, "Something went wrong"
      rescue RuntimeError => e
      end
      logger.notify e

      log_hash = expected_hash({
        short_message: "#{e.class}: #{e.message}",
        level: GELF::ERROR,
        full_message: "Backtrace:\n#{e.backtrace.join("\n")}"
      })
      expect(GELF::Notifier.last_hash).to eq(log_hash)
    end
  end

  describe "#from_config" do
    it "must accept the logger tree from a config file" do
      config = {
        :level               => "INFO",
        :'udp.host'          => "127.0.0.2",
        :'udp.port'          => 12345,
        :'gelf.facility'     => "my_facility",
        :'gelf.maxChunkSize' => 8100,
        :'console.enabled'   => true
      }
      logger = described_class.from_config(config)
      
      expect(logger.host).to eq(config[:'udp.host'])
      expect(logger.port).to eq(config[:'udp.port'])
      expect(logger.default_options["facility"]).to eq(config[:'gelf.facility'])
      expect(logger.max_chunk_size).to eq(config[:'gelf.maxChunkSize'])
      expect(logger.level).to eq(GELF.const_get(config[:level].upcase))
      expect(logger.instance_variable_get(:'@stdout_logger')).to_not be_nil
    end

    it "must raise an ArgumentError if any settings are invalid, listing problem values" do
      config = {
        :level               => "INFO",
        :'udp.host'          => "127.0.0.2",
        :'udp.port'          => 12345,
        :'gelf.facility'     => "my_facility",
        :'gelf.maxChunkSize' => 8100
      }

      %i{udp.host udp.port gelf.facility gelf.maxChunkSize}.each do |invalid_config|
        begin
          partially_invalid_config = config.dup
          partially_invalid_config[invalid_config] = nil
          described_class.from_config(partially_invalid_config)
          expect("this").to_not be("reached"), "#from_config didn't raise an ArgumentError"
        rescue ArgumentError => e
          expect(e.message).to include(invalid_config.to_s)
        end
      end
    end
  end
end

spec/extra_hash_methods_spec.rb
require "blinkbox/utilities/extra_hash_methods"

context Blinkbox::ExtraHashMethods do
  describe "#flatten_hash" do
    it "must reduce a two level deep hash" do
      hash = {
        'firstA' => {
          'second' => "firstA.second"
        },
        'firstB' => "firstB"
      }

      output = hash.extend(described_class).shallow!

      expect(output).to eql(true)
      hash.each do |key, value|
        expect(key).to eq(value)
      end
    end

    it "must reduce a two level deep hash with symbol keys" do
      hash = {
        firstA: {
          second: "firstA.second"
        },
        firstB: :firstB
      }

      output = hash.extend(described_class).shallow!

      expect(output).to eql(true)
      hash.each do |key, value|
        expect(key).to eq(value)
      end
    end

    it "must not change a shallow hash" do
      hash = {
        'firstA' => "firstA",
        'firstB' => "firstB"
      }

      output = hash.extend(described_class).shallow!

      expect(output).to eql(false)
      hash.each do |key, value|
        expect(key).to eq(value)
      end
    end

    it "must allow joining with other strings" do
      hash = {
        'firstA' => {
          'second' => "firstA-second"
        },
        'firstB' => "firstB"
      }

      output = hash.extend(described_class).shallow!('-')

      expect(output).to eql(true)
      hash.each do |key, value|
        expect(key).to eq(value)
      end
    end

    it "must output clashing keys" do
      hash = {
        'firstA' => {
          'second' => "firstA.second"
        },
        'firstA.second' => "overwritten"
      }

      output = hash.extend(described_class).shallow!
      expect(output).to eql(["firstA.second"])
    end

    it "must reduce a three level deep hash" do
      hash = {
        'firstA' => {
          'secondA' => {
            'third' => "firstA.secondA.third"
          },
          'secondB' => "firstA.secondB"
        },
        'firstB' => "firstB"
      }

      expected = {
        'firstA.secondA.third' => "firstA.secondA.third",
        'firstA.secondB' => "firstA.secondB",
        'firstB' => "firstB"
      }

      output = hash.extend(described_class).shallow!

      expect(hash).to eql(expected)
    end
  end
end

lib/blinkbox/common_logging/logger.rb
require "gelf"
require "logger"

module Blinkbox
  class CommonLogging < GELF::Logger
    def initialize(host: "localhost", port: 12201, facility: $0, max_size: 8192, facility_version: nil, echo_to_console: false)
      validity_issues = []
      validity_issues.push("host") unless host.is_a?(String)
      validity_issues.push("port") unless port.is_a?(Integer)
      validity_issues.push("facility") unless facility.is_a?(String)
      validity_issues.push("max_size") unless max_size.is_a?(Integer)
      raise ArgumentError, "Cannot start the logger, the following settings weren't valid: #{validity_issues.join(", ")}" if !validity_issues.empty?

      if echo_to_console
        @stdout_logger = Logger.new(STDOUT)
        @stdout_logger.formatter = proc do |severity, datetime, progname, msg|
          # Using console output is only for live-debug, so we can strip the timestamp, progname
          l = msg.dup
          l.delete("facility")
          l.delete(:facility)
          log_msg = "#{severity[0..0]} #{l.delete("short_message")}#{l.delete(:short_message)}"
          log_msg << " (#{l})" if l.any?
          log_msg << "\n"
          log_msg
        end
      end
      options = { facility: facility }
      options[:facilityVersion] = facility_version unless facility_version.nil?
      super(host, port, max_size, options)
    end

    # Allows the setting of the facility version default parameter.
    # @param [String] semver The semantic version of the service which is sending logs
    def facility_version=(semver)
      self.default_options['facilityVersion'] = semver
    end

    # Accepts a hash detailing log config settings in the format used in blinkbox Books
    # config files and returns a CommonLogging object.
    #
    # @example Load from config
    #   config = Blinkbox::CommonConfig.new
    #   logger = Blinkbox::CommonLogging.from_config(config.tree(:logging))
    #
    # @param [Hash] hash The returned hash from CommonConfig#tree
    # @option hash [String] :udp.host (localhost) The Graylog host
    # @option hash [Integer] :udp.port (12201) The Graylog port number
    # @option hash [String] :gelf.facility Organisation/app_name eg. Marvin/cover_processor
    # @option hash [Integer] :gelf.maxChunkSize (8192) The default size for UDP chunks sent to the server.
    # @option hash ["DEBUG","INFO","WARN","ERROR","FATAL"] :level ("INFO") The level below which messages will not be sent to Graylog.
    # @return [CommonLogging]
    def self.from_config(hash)
      mapping = {
        host:            :'udp.host',
        port:            :'udp.port',
        facility:        :'gelf.facility',
        max_size:        :'gelf.maxChunkSize',
        echo_to_console: :'console.enabled'
      }
      begin
        logger = new(Hash[mapping.map { |k, v| [k, hash[v]] }])
      rescue ArgumentError => e
        msg = e.message
        mapping.each do |k, v|
          msg.sub!(k.to_s, v.to_s)
        end
        raise e.class, msg
      end
      logger.level = GELF.const_get(hash[:level].upcase) rescue GELF::INFO
      logger
    end

    private
    def notify_with_level!(message_level, msg)
      if @stdout_logger
        level = GELF::Levels.constants[message_level].to_s.downcase.to_sym
        @stdout_logger.send(level, msg)
      end
      msg.extend(ExtraHashMethods).shallow! if msg.is_a?(Hash)
      super(message_level, msg)
    end
  end
end

lib/blinkbox/utilities/extra_hash_methods.rb
module Blinkbox
  module ExtraHashMethods
    # Turns a deep/nested hash into a shallow one by concatenating nested key names.
    #
    # @example: Removing nesting
    #   hash = {
    #     "a" => {
    #       "1" => 'will be a.1'
    #     },
    #     "b" => 'will be b'
    #   }
    #   hash.extend(ExtraHashMethods).shallow!
    #   # => true
    #   p hash
    #   # => { "a.1" => 'will be a.1', "b" => 'will be b' }
    #
    # @example: Removing nesting - custom join string
    #   hash = {
    #     "a" => {
    #       "1" => 'will be merged down'
    #     },
    #     "b" => 'will be the same'
    #   }
    #   hash.extend(ExtraHashMethods).shallow!("~")
    #   # => true
    #   p hash
    #   # => { "a~1" => 'will be merged down', "b" => 'will be the same' }
    #
    # @example: No removal needed
    #   hash = {
    #     "a" => 'no deep keys here'
    #   }
    #   hash.extend(ExtraHashMethods).shallow!
    #   # => false
    #   p hash
    #   # => { "a" => 'no deep keys here' }
    #
    # @example: Conflicting key names
    #   hash = {
    #     "a" => {
    #       "1" => 'will be a.1'
    #     },
    #     "a.1" => 'will be lost!'
    #   }
    #   hash.extend(ExtraHashMethods).shallow!
    #   # => ["a.1"]
    #   p hash
    #   # => { "a.1" => 'will be a.1' }
    #
    # @param [String] join The string to use to join keys.
    # @return [Boolean] Whether any keys were changed
    def shallow!(join = ".")
      shallowed = false
      keys = self.keys.select { |k| self[k].is_a?(Hash) }
      while (k = keys.shift) do
        shallowed = shallowed || true
        v = self.delete(k)
        v.each do |sub_k, sub_v|
          new_key = [k, sub_k].join(join)
          if self.has_key?(new_key)
            shallowed = [] unless shallowed.is_a?(Array)
            shallowed.push(new_key)
          end
          self[new_key] = if sub_v.is_a?(Hash)
            sub_v.extend(ExtraHashMethods).shallow!(join)
            keys.push(new_key)
            sub_v
          else
            sub_v
          end
        end
      end
      shallowed
    end
  end
end

lib/blinkbox/common_logging.rb
require "blinkbox/utilities/extra_hash_methods"
require "gelf"
require "logger"

module Blinkbox
  class CommonLogging < GELF::Logger
    def initialize(host: "localhost", port: 12201, facility: $0, max_size: 8192, facility_version: nil, echo_to_console: false)
      validity_issues = []
      validity_issues.push("host") unless host.is_a?(String)
      validity_issues.push("port") unless port.is_a?(Integer)
      validity_issues.push("facility") unless facility.is_a?(String)
      validity_issues.push("max_size") unless max_size.is_a?(Integer)
      raise ArgumentError, "Cannot start the logger, the following settings weren't valid: #{validity_issues.join(", ")}" if !validity_issues.empty?

      if echo_to_console
        @stdout_logger = Logger.new(STDOUT)
        @stdout_logger.formatter = proc do |severity, datetime, progname, msg|
          # Using console output is only for live-debug, so we can strip the timestamp, progname
          l = msg.dup
          l.delete("facility")
          l.delete(:facility)
          log_msg = "#{severity[0..0]} #{l.delete("short_message")}#{l.delete(:short_message)}"
          log_msg << " (#{l})" if l.any?
          log_msg << "\n"
          log_msg
        end
      end
      options = { facility: facility }
      options[:facilityVersion] = facility_version unless facility_version.nil?
      super(host, port, max_size, options)
    end

    # Allows the setting of the facility version default parameter.
    # @param [String] semver The semantic version of the service which is sending logs
    def facility_version=(semver)
      self.default_options['facilityVersion'] = semver
    end

    # Accepts a hash detailing log config settings in the format used in blinkbox Books
    # config files and returns a CommonLogging object.
    #
    # @example Load from config
    #   config = Blinkbox::CommonConfig.new
    #   logger = Blinkbox::CommonLogging.from_config(config.tree(:logging))
    #
    # @param [Hash] hash The returned hash from CommonConfig#tree
    # @option hash [String] :udp.host (localhost) The Graylog host
    # @option hash [Integer] :udp.port (12201) The Graylog port number
    # @option hash [String] :gelf.facility Organisation/app_name eg. Marvin/cover_processor
    # @option hash [Integer] :gelf.maxChunkSize (8192) The default size for UDP chunks sent to the server.
    # @option hash ["DEBUG","INFO","WARN","ERROR","FATAL"] :level ("INFO") The level below which messages will not be sent to Graylog.
    # @return [CommonLogging]
    def self.from_config(hash)
      mapping = {
        host:            :'udp.host',
        port:            :'udp.port',
        facility:        :'gelf.facility',
        max_size:        :'gelf.maxChunkSize',
        echo_to_console: :'console.enabled'
      }
      begin
        logger = new(Hash[mapping.map { |k, v| [k, hash[v]] }])
      rescue ArgumentError => e
        msg = e.message
        mapping.each do |k, v|
          msg.sub!(k.to_s, v.to_s)
        end
        raise e.class, msg
      end
      logger.level = GELF.const_get(hash[:level].upcase) rescue GELF::INFO
      logger.level_mapping = :direct
      logger
    end

    private
    def notify_with_level!(message_level, msg)
      if @stdout_logger
        level = GELF::Levels.constants[message_level].to_s.downcase.to_sym
        @stdout_logger.send(level, msg)
      end
      msg.extend(ExtraHashMethods).shallow! if msg.is_a?(Hash)
      begin
        super(message_level, msg)
      rescue SocketError
        # Cannot send data to unknown host, ignoring
      end
    end
  end
end
